secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/SecurePreferences.java

```diff
@@ -24,52 +24,94 @@
+            } else {
+                Log.e(SecurePreferences.class.getName(),
+                        context.getString(R.string.message_problem_encryption));
+            }
         } else {
-            Log.e(SecurePreferences.class.getName(),
-                    context.getString(R.string.message_problem_encryption));
+            byte[] saltBytes = KeystoreTool.calculateSalt();
+            String salt;
+            try {
+                if (saltBytes != null) {
+                    salt = new String(saltBytes, 0, saltBytes.length, KEY_CHARSET);
+                } else {
+                    Log.e(SecurePreferences.class.getName(),
+                            context.getString(R.string.message_problem_salt_creation));
+                    return;
+                }
+            } catch (UnsupportedEncodingException e) {
+                if (BuildConfig.DEBUG) {
+                    Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
+                }
+                return;
+            }
+            String hashedValue = KeystoreTool.getSHA512(value, salt);
+            if (hashedValue != null) {
+                setSecureValue(key, hashedValue, context);
+                setSecureValue(SALT_PREFIX + key, salt, context);
+            } else {
+                Log.e(SecurePreferences.class.getName(),
+                        context.getString(R.string.message_problem_hashing));
+            }
         }
     }
 
     @Nullable
-    public static String getValue(@NonNull String key, @NonNull Context context) {
+    public static String getValue(@NonNull String key,
+                                  @NonNull Context context,
+                                  @NonNull SecureMethod secureMethod) {
+
+        if (secureMethod.equals(SecureMethod.METHOD_ENCRYPT)) {
+            String result = getSecureValue(key, context);
+            try {
+                return KeystoreTool.decryptMessage(context, result != null
+                        ? result : context.getString(R.string.message_nothing_found));
+            } catch (NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException |
+                    UnrecoverableEntryException | IOException | CertificateException |
+                    InvalidKeyException | KeyStoreException e) {
 
-        String result = getSecureValue(key, context);
-        try {
-            return KeystoreTool.decryptMessage(context, result != null
-                    ? result : context.getString(R.string.message_nothing_found));
-        } catch (NoSuchPaddingException | NoSuchAlgorithmException | NoSuchProviderException |
-                UnrecoverableEntryException | IOException | CertificateException |
-                InvalidKeyException | KeyStoreException e) {
-
-            Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
-            return null;
+                Log.e(SecurePreferences.class.getName(), e.getMessage(), e);
+                return null;
```

Comment:
I think it's a good idea to throw a custom RuntimeException here so that the developer can get feedback that it didn't work